runtime/src/iree/base/tree.h

```diff
@@ -0,0 +1,82 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_BASE_TREE_H
+#define IREE_BASE_TREE_H
+
+#include "iree/base/allocator.h"
+#include "iree/base/config.h"
+#include "stdint.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+typedef struct iree_tree_node_t iree_tree_node_t;
+typedef struct iree_tree_t iree_tree_t;
+
+typedef enum iree_tree_walk_type_e {
+  IREE_TREE_WALK_PREORDER,
+  IREE_TREE_WALK_INORDER,
+  IREE_TREE_WALK_POSTORDER,
+} iree_tree_walk_type_e;
+
+typedef struct iree_tree_node_t iree_tree_node_t;
+typedef struct iree_tree_node_t {
+  bool red;
+  iree_tree_node_t* left;
+  iree_tree_node_t* right;
+  iree_tree_node_t* parent;
+  iree_host_size_t key;
+  bool is_sentinel;
+  uint8_t* data;
+} iree_tree_node_t;
+
+typedef struct iree_tree_t {
+  uint32_t element_size;
+  iree_allocator_t allocator;
+  iree_tree_node_t* root;
+  iree_host_size_t size;
+  iree_tree_node_t* cache;  // Cache for deleted nodes
+  iree_tree_node_t nil;
+  uint8_t* initial_node_cache;
+  iree_host_size_t initial_node_cache_size;
+} iree_tree_t;
+
+void* iree_tree_node_get_data(iree_tree_node_t* node);
+iree_host_size_t iree_tree_node_get_key(iree_tree_node_t* node);
+
+iree_status_t iree_tree_initialize(iree_allocator_t allocator,
+                                   iree_host_size_t element_size,
+                                   void* initial_node_cache,
+                                   iree_host_size_t initial_node_cache_size,
+                                   iree_tree_t* tree);
+uint32_t iree_tree_get_data_size(iree_tree_t* tree);
```

Comment:
read-only accessors that don't mutate state should have the name of the field (`iree_tree_data_size`) and setters should have set_ (`iree_tree_set_data_size`)
https://google.github.io/styleguide/cppguide.html#Function_Names

in non-accessor cases another word than `get` is often useful for denoting function behavior to users - a `get` implies an O(1) kind of operation while a `lookup`/`find` may be more expensive as it's searching a data structure - a reader can usually then know that calling a `get`/accessor a few times is going to be cheap but that it's worth it to cache a `lookup`/`find` in a local (lookups and finds usually return statuses while a get would not be expected to)